src/correlation.rs

```diff
@@ -198,6 +198,24 @@ impl Correlations {
 
         Ok(())
     }
+
+    pub async fn delete_for_user(&self, user_id: &str) -> Result<(), CorrelationError> {
+        // Cautious to avoid deadlock with read and write.
+        let correlation_ids: Vec<String> = {
+            let correlations = self.0.read().await;
+            // Finding all correlations for user_id
+            correlations
+                .iter()
+                .filter(|corr_map| corr_map.1.user_id.as_str() == user_id)
+                .map(|x| x.0.clone())
+                .collect()
+        };
+        // Removing them from the hashmap.
+        for id in correlation_ids.iter() {
+            self.0.write().await.remove_entry(id);
+        }
```

Comment:
Wouldn't it be a good idea to get a write lock and perform this operation in one go?